### DISCUSSION TITLE
Example for mp3/Opus encoding

### DISCUSSION THREAD

**Discussion by massa007** [pschatzmann/arduino-audio-tools#1016] [Q&A] [answered]
Hi all,

I am struggling with the mp3 encoding.

I already tried including the LAME codec and replacing the new WAVEncoder() with new MP3EncoderLAME() but PlatformIO returns an error:

` EncodedAudioStream out(&file, new MP3EncoderLAME());`
![Bildschirmfoto 2023-10-09 um 18 22 54](https://github.com/pschatzmann/arduino-audio-tools/assets/16576171/28fa0631-560e-46a9-8215-ec15006a3a0f)

Sorry for German language, but it says that an object of type MP3EncoderLAME of an abstract class is not allowed.
Can anyone provide a similar example like for wav?

Thanks a lot for every help 

PS: This is the example I use as orientation.
```
#include "AudioTools.h"
#include "SdFat.h"

uint16_t sample_rate = 44100;
uint8_t channels = 2;                                             // The stream will have 2 channels 
NoiseGenerator<int16_t> noise(32000);                             // subclass of SoundGenerator with max amplitude of 32000
GeneratedSoundStream<int16_t> in_stream(noise);                   // Stream generated from sine wave
SdFat SD;
File audioFile;                                                   // final output stream
EncodedAudioStream out_stream(&audioFile, new WAVEncoder());             // encode as wav file
StreamCopy copier(out_stream, in_stream);                                // copies sound to out

void setup(){
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);  

  auto cfg = noise.defaultConfig();
  cfg.sample_rate = sample_rate;
  cfg.channels = channels;
  cfg.bits_per_sample = 16;
  noise.begin(cfg);
  in_stream.begin();

  // we need to provide the audio information to the encoder
  out_stream.begin(cfg);
  // open the output file
  SD.begin(SdSpiConfig(PIN_CS, DEDICATED_SPI, SD_SCK_MHZ(2)));
  audioFile = SD.open("/test/002.wav", O_WRITE | O_CREAT);
}

void loop(){
    copier.copy();  
    // audioFile.flush(); // force write down of data
}
```

PPS:
I also tried to use the Oous Codec but get the following error:
![grafik](https://github.com/pschatzmann/arduino-audio-tools/assets/16576171/e070bb39-3beb-46f0-9a42-80cd41204eeb)

Would it be possible to include more examples if the usage of different codecs is not similar? Probably I am just stupid but I don´t know how to proceed!

**Comment by pschatzmann** [comment, marked as answer]
I guess you are missing the necessary includes in your sketch! 
Check the Codec Wiki documentation: It tells you the class name and the include

**Comment by massa007** [comment]
It is definitely installd and included:
![grafik](https://github.com/pschatzmann/arduino-audio-tools/assets/16576171/e7d4d129-ff4a-4e52-9a79-a2afd2943704)

Any other idea?

**Comment by pschatzmann** [comment]
Did you try to compile the [codec test sketches](https://github.com/pschatzmann/arduino-audio-tools/tree/main/examples/tests/codecs) ?

**Comment by pschatzmann** [comment]
Since you did not provide any test sketch for me, I created a very simple one. 
The following compiles w/o error:
```C++
#include "AudioTools.h"
#include "AudioCodecs/CodecMP3LAME.h"

I2SStream out; 
EncodedAudioStream decoder(&out, new MP3EncoderLAME()); // encode and write

void setup() {
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Info);
}

// copy the data
void loop() {
}
```

**Comment by massa007** [comment]
Just did and it works with that example! So it seems like I have to declare it explicitly:
![grafik](https://github.com/pschatzmann/arduino-audio-tools/assets/16576171/3e0f3e91-ef56-4638-ac24-bc55a66d5be5)

Compilation was succesful! 

Cannot thank you enough for your awesome support and work here!!

**Comment by pschatzmann** [comment]
That's strange: If you check in the Board Manager: what ESP32 release are you using ?

**Reply by massa007** [reply]
Is this what you are looking for? Not sure about the differences in Arduino IDE and PlatformIO..

![grafik](https://github.com/pschatzmann/arduino-audio-tools/assets/16576171/2577b4cd-3866-410b-ad3a-a1a87d9db26d)

**Comment by pschatzmann** [comment]
Maybe you should use some more recent version: The current Arduino Version is 2.0.14 and if you are still using 2.0.5 you are quite some versions behind.

If you check with https://github.com/platformio/platform-espressif32 the actual version is 6.4.0

**Reply by massa007** [reply]
Good hint, thank you! Just switched to the most current version but it still behaves the same.
But there might be still an issue as playback worked with raw and mp3 for example but now that I changed to opus I get weird noise and the following messages:

```
[I] StreamCopy.h : 71 - buffer_size=1024
[E] CodecOpus.h : 195 - opus_decode: corrupted stream
[I] StreamCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[E] CodecOpus.h : 195 - opus_decode: corrupted stream
[I] StreamCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[E] CodecOpus.h : 195 - opus_decode: corrupted stream
[I] StreamCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
[E] CodecOpus.h : 195 - opus_decode: corrupted stream
[I] StreamCopy.h : 136 - StreamCopy::copy 1024 -> 1024 -> 1024 bytes - in 1 hops
```

Opus file played well with vlc on my macbook!

**Comment by pschatzmann** [comment]
Opus needs to be used with a container, this is usually ogg.
Otherwise you need to make sure that you process full opus segments via another mechanism.

My recommendation would be to use ADPC. It is lean, fast, efficient on microcontrollers and it is compressing...
